### PR TITLE
[WIP] perf improvements for strftime

### LINKED ISSUE #44764: PERF: strftime is slow
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this issue exists on the master branch of pandas.

### Reproducible Example

I found pd.DatatimeIndex.strftime is pretty slow when data is large.

In the following I made a simple benchmark. `method_b` first stores 'year', 'month', 'day', 'hour', 'minute', 'second', then convert them to string with f-formatter. Although it's written in python, the time spent is significantly lower.

```python
import time
import numpy as np
import pandas as pd
from joblib import Parallel, delayed

def timer(f):
    def inner(*args, **kwargs):
        s = time.time()
        result = f(*args, **kwargs)
        e = time.time()
        return e - s
    return inner


@timer
def method_a(index):
    return index.strftime("%Y-%m-%d %H:%M:%S")

@timer
def method_b(index):
    attrs = ('year', 'month', 'day', 'hour', 'minute', 'second')
    parts = [getattr(index, at) for at in attrs]
    b = []
    for year, month, day, hour, minute, second in zip(*parts):
        b.append(f'{year}-{month:02}-{day:02} {hour:02}:{minute:02}:{second:02}')
    b = pd.Index(b)
    return b

index = pd.date_range('2000', '2020', freq='1min')

@delayed
def profile(p):
    n = int(10 ** p)
    time_a = method_a(index[:n])
    time_b = method_b(index[:n])
    return n, time_a, time_b

records = Parallel(10, verbose=10)(profile(p) for p in np.arange(1, 7.1, 0.1))
```
```python
pd.DataFrame(records, columns=['n', 'time_a', 'time_b']).set_index('n').plot(figsize=(10, 8))
```
image

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.8.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.8.0-63-generic
Version          : #71-Ubuntu SMP Tue Jul 13 15:59:12 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.4
numpy            : 1.20.0
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 58.0.4
Cython           : 0.29.24
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.2
html5lib         : 1.1
pymysql          : 0.9.3
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : 2.11.3
IPython          : 7.29.0
pandas_datareader: 0.9.0
bs4              : None
bottleneck       : 1.3.2
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : 5.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : 1.4.22
tables           : None
tabulate         : 0.8.9
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.54.1

</details>

### Prior Performance

_No response_

### LINKED ISSUE #46252: BUG: Period formatting directive %l (millisecond) and %u (microseconds) provide wrong results
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

p = pd.period_range("2003-01-01 12:01:01.123456789", periods=2, freq="n")
```

### Issue Description

`p[0].strftime("millis: %l - micros: %u - nanos: %n")`

yields `'millis: 789 - micros: 456789 - nanos: 123456789'`

### Expected Behavior

THe correct result should be `'millis: 123 - micros: 123456 - nanos: 123456789'`

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 2c23f48ef23d650ae62642ca5f192b6ed0fe986f
python           : 3.8.11.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19043
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : fr_FR.cp1252
pandas           : 1.5.0.dev0+420.g2c23f48ef2.dirty
numpy            : 1.20.3
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.0.1
setuptools       : 58.0.4
Cython           : 0.29.24
pytest           : 6.2.5
hypothesis       : 6.29.3
sphinx           : 4.2.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.27.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
markupsafe       : 2.0.1
matplotlib       : 3.4.2
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### LINKED ISSUE #46319: BUG: `UnicodeError` when using `Period.strftime` with non-utf8 locale
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import locale

locale.setlocale(locale.LC_ALL, "zh_CN")
p = pd.Period("2018-03-11 13:00", freq="H")
print(p.strftime("%p"))  # UnicodeDecodeError: 'utf-8' codec can't decode byte 0xcf ...
```

### Issue Description

`Period.strftime("%p")` prints the locale-specific version of AM or PM. When the locale uses a non-utf8 compliant encoding, it crashes.

This bug does not happen with others, for example `Timestamp.strftime`.

### Expected Behavior

No error, printing the actual string representing AM or PM

### Installed Versions

This is on the `main` branch head

### PR DESCRIPTION
**Pull request by smarie**
Changelog

 - Added the default format "%Y-%m-%d %H:%M:%S" as basic format (same as format=None), as suggested by @auderson.

- [ ] closes #44764 , closes #46252, closes #46319
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.
- [ ] Post `timeit` results showing the perf improvements

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Can you post %timeit results showing the perf improvement?  (edit OP)

**Comment by smarie** [comment]
Sure. I edited the todo list accordingly.

**Comment by smarie** [comment]
Note: there seem to be a related issue in #29461 , solved by #34668 . Still I will favor the current road where I'll modify the `format_array_from_datetime` function only.

**Review by phofl — COMMENTED** [review]
Are these issues all connected? In general we prefer if one pr tackels one topic, not multiple at once

**Comment by mroeschke** [comment]
> Are these issues all connected? In general we prefer if one pr tackels one topic, not multiple at once

Agreed, with the current diff, this PR is likely too large for someone to review in earnest.

**Comment by smarie** [comment]
Noted - I'll at least try to move the two fixes for `Period` into separate PRs, before this final big one about `fast_strftime` is reviewed.

**Comment by smarie** [comment]
Quick status on this PR: today I reached a point where all CI checks pass (finally !!!), and therefore I was able to try reproducing the benchmark in #44764 . Results are pretty good !

image

<details>
```python
import time
import numpy as np
import pandas as pd
from joblib import Parallel, delayed


def timer(f):
    def inner(*args, **kwargs):
        s = time.time()
        result = f(*args, **kwargs)
        e = time.time()
        return e - s
    return inner


@timer
def method_legacy(index):
    return index.strftime("%Y-%m-%d_%H:%M:%S", fast_strftime=False)


@timer
def method_manual(index):
    attrs = ('year', 'month', 'day', 'hour', 'minute', 'second')
    parts = [getattr(index, at) for at in attrs]
    b = []
    for year, month, day, hour, minute, second in zip(*parts):
        b.append(f'{year}-{month:02}-{day:02}_{hour:02}:{minute:02}:{second:02}')
    b = pd.Index(b)
    return b


@timer
def method_new_and_legacy_default(index):
    return index.strftime(None)


@timer
def method_new_custom(index):
    return index.strftime("%Y-%m-%d_%H:%M:%S")


@timer
def method_new_basic_formats(index):
    return index.strftime("%Y-%m-%d %H:%M:%S")


index = pd.date_range('2000', '2020', freq='1min')


@delayed
def profile(p):
    n = int(10 ** p)
    time_legacy = method_legacy(index[:n])
    time_manual = method_manual(index[:n])
    time_new_and_legacy_default = method_new_and_legacy_default(index[:n])
    time_new_custom = method_new_custom(index[:n])
    time_new_basic_formats = method_new_basic_formats(index[:n])
    return n, time_legacy, time_manual, time_new_and_legacy_default, \
           time_new_custom, time_new_basic_formats


records = Parallel(10, verbose=10, max_nbytes='500M')(
    profile(p) for p in np.arange(1, 7.1, 0.1)
)


pd.DataFrame(records, columns=['n', 'time_legacy', 'time_manual',
                               'time_new_and_legacy_default', 'time_new_custom',
                               'time_new_basic_formats'
                               ]).set_index(
    'n').plot(figsize=(10, 8))

import matplotlib.pyplot as plt
plt.show()
```
</details>

**Comment by jreback** [comment]
@smarie thins way too large a PR

pls split this up into logical pieces

**Comment by smarie** [comment]
@jreback yes I was going to split it as explained above, but I wanted first to reach a "clean" state where CI is happy and the benchmark shows good results. Now I will try (next week) to split this into pieces, but as already mentioned, there is not much I can do here: `strftime` was used in many places (`Period`, `Timestamp`, csv formatter, excel formatter, sqlite adapter, some `repr` methods, etc.) and all of them needed replacement. I will do my best do separate between

 - bug fixes
 - quick fixes for all repr and the like (not requiring core changes)
 - the "core" replacement for datetime arrays
 - the same for periods
 - ... maybe other similar

**Comment by auderson** [comment]
This is impressive😀

File: pandas/_libs/tslibs/period.pyx [hunk 1]
```diff
@@ -2340,14 +2392,70 @@ cdef class _Period(PeriodMixin):
         object_state = None, self.freq, self.ordinal
         return (Period, object_state)
 
+    def fast_strftime(self, fmt_str: str, loc_s: object) -> str:
+        """A faster alternative to `strftime` using string formatting.
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/period.pyx:2396]: nitpick newline after the """

File: pandas/_libs/tslib.pyx [hunk 2]
```diff
@@ -101,7 +105,8 @@ def format_array_from_datetime(
     ndarray[int64_t] values,
```
Comment (jbrockmendel) [inline pandas/_libs/tslib.pyx:105]: if a tslibs.strftime does get implemented, that might be a good home for format_array_from_datetime

**Comment by smarie** [comment]
bot: please leave this open. #46759 and #46405 are the next bits of this PR to consider

**Comment by Paul-AUB** [comment]
Hello @smarie, it seems there is a flake8 issue (`pandas/tests/io/formats/test_format.py:3253:5: F811 redefinition of unused 'test_period_custom' from line 3201`) to fix and a conflict to solve before this can be merged. Could you do that please ? Feature is really interesting !

**Comment by smarie** [comment]
Thanks @Paul-AUB ! Actually this PR is just here as a reference, it is split in several pieces. https://github.com/pandas-dev/pandas/pull/46759 and https://github.com/pandas-dev/pandas/pull/46405 are the next bits of this PR to consider, the first is actually quite close to be merged. The second has a side effect in the CI concerning some locale-dependent setting

**Comment by smarie** [comment]
Hi @mroeschke, yes I plan to finish this now that most of its "other aspects" have been merged (through other PRs #46361, #46759, #46405, #47570). I'll try to find some time later this month

**Pull request closed without merging**